### PR TITLE
Issue 125 model well-posedness

### LINKED ISSUE #125: Model well-posedness
When checking that a PDE model is well-posed, make sure that all variables that appear in the rhs dict values expression trees (i.e. the model equations) also appear in the rhs keys. 
For DAE models, give some "free passes" corresponding to the number of algebraic equations (for which we don't explicitly define a variable)

Most submodels won't be well-posed, but they don't need to be

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

Fixes #125 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by tinosulzer** [comment]
This will be affected by the changes in #128

File: tests/test_models/test_base_model.py [hunk 1]
```diff
@@ -156,7 +156,47 @@ def test_update(self):
...
+        model.rhs = {c: 5 * pybamm.div(pybamm.grad(d)) - 1}
+        model.algebraic = [e - c - d]
+        with self.assertRaises(pybamm.ModelError) as error:
+            model.check_well_posedness()
+            self.assertIn("underdetermined", error.exception.args[0])
+
+        # Underdetermined model - not enough algebraic equations
+        model.rhs = {c: 5 * pybamm.div(pybamm.grad(d)) - 1, d: -c - e}
+        model.algebraic = []
+        with self.assertRaises(pybamm.ModelError) as error:
```
Comment (martinjrobins) [inline tests/test_models/test_base_model.py]: could also use assertRaisesRegex here, eg
```python
with self.assertRaisesRegex(pybamm.ModelError, 'underdetermined'):
   model.check_well_posedness())
```

File: tests/test_models/test_base_model.py [hunk 2]
```diff
@@ -156,7 +156,47 @@ def test_update(self):
...
+            model.check_well_posedness()
+            self.assertIn("underdetermined", error.exception.args[0])
+
+        # Overdetermined model - too many algebraic equations
+        # Model cannot be overdetermined by too many differential equations, as rhs
+        # keys are all unique variables
+        model.algebraic = [c - d, e + d]
+        with self.assertRaises(pybamm.ModelError) as error:
+            model.check_well_posedness()
+        self.assertIn("overdetermined", error.exception.args[0])
```
Comment (martinjrobins) [inline tests/test_models/test_base_model.py]: should be indented?, or see previous comment

File: tests/test_models/test_base_model.py (hunk 1 above)
Comment (tinosulzer) [reply tests/test_models/test_base_model.py]: cool, cheers, didn't know about assertRaisesRegex

**Pull request merged**